Node wrapper

humphd opened the pull request:
Created node.js code to use this easily, with tests and docs.

humphd commented:
I'll need to get some help with versioning for CI to pass (which version of Go tooling to use?), but this is working locally.

tarasglek commented:
needs latest go for wasm

tarasglek commented on eget.js:
```diff
@@ -0,0 +1,343 @@
+ * @param {string} errorStr - Error string from eget WASM
+ * @returns {string|null} Extracted URL or null if not found
+ */
+function parseErrorForUrl(errorStr) {
```
I purposely made errors be json so you wouldn't have to regexp stuff :)

tarasglek commented on eget.js:
```diff
@@ -0,0 +1,343 @@
+    while (attempts < maxAttempts) {
+      const result = await this.run(args);
+
+      if (result.success) {
```
maybe for a followup, but mac/linux will need chmod +x on the resulting files

humphd replied:
I *think* I've done this correctly, but it's kind of tricky, since eget can download more than just binaries.

tarasglek commented on eget.js:
```diff
@@ -0,0 +1,343 @@
+
+    let module, instance;
+    try {
+      module = await WebAssembly.compile(wasmBytes);
```
I suspect you wanna cache module in a global and not re-read-recompile  every retry

tarasglek commented on eget.js:
```diff
@@ -0,0 +1,343 @@
+ * Detects the current system platform and architecture as expected by eget.
+ * @returns {string} System string in format 'platform/arch' (e.g., 'linux/amd64')
+ */
+export function detectSystem() {
```
nice

humphd commented:
Did some fixes based on your review.  This still needs work before I can land it, but it's getting closer.

tarasglek commented:
There is also a small security issue, we need to ensure that url->path translation doesn't escape http/ prefix...by making dir absolute and ensuring that it has the expected prefix. I think that's better done on go side when converting url to path.

In practice we are safe cos urls are all returned by github, but we might integrate patches for eget that  support gitlab, etc

humphd commented:
@tarasglek I've done some more work on this, but I'm now running into a hang in the WASM execution itself.  Here's a reduced test case:

```js
import { eget } from "./eget.js";

async function main() {
  await eget('getsops/sops', { verbose: true });
  console.log('done')
}

main();
```

Here's what happens:

```sh
$ node test.js
[eget.wasm] Running eget with args: --system darwin/arm64 getsops/sops
(node:66525) ExperimentalWarning: WASI is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
[eget.wasm] using stderr file /var/folders/82/9ymnd56s1g17vffjb280x32c0000gn/T/eget.stderr_de54fc51-f65e-48b8-9265-10cb8442bbac.txt
[eget.wasm] Compiled WASM module: /Users/humphd/repos/eget.wasm/eget.wasm
[eget.wasm] Starting WASM execution...
[eget.wasm] WASM execution finished with exit code: 1
[eget.wasm] exited with code 1: open /tmp/https/api.github.com/repos/getsops/sops/releases/latest: No such file or directory
[eget.wasm] Attempt 1: eget failed - open /tmp/https/api.github.com/repos/getsops/sops/releases/latest: No such file or directory
[eget.wasm] Downloading: https://api.github.com/repos/getsops/sops/releases/latest to /Users/humphd/repos/eget.wasm/tmp/https/api.github.com/repos/getsops/sops/releases/latest
[eget.wasm] Saved to: /Users/humphd/repos/eget.wasm/tmp/https/api.github.com/repos/getsops/sops/releases/latest
[eget.wasm] using stderr file /var/folders/82/9ymnd56s1g17vffjb280x32c0000gn/T/eget.stderr_6b7a2aa7-784e-4578-a397-1f8e20864f0e.txt
[eget.wasm] Compiled WASM module: /Users/humphd/repos/eget.wasm/eget.wasm
[eget.wasm] Starting WASM execution...
```

It then hangs forever.  So the first `eget` call fails (nothing in filesystem, which is expected); it downloads the file to the filesystem (I see it); but then it doesn't continue to parse the JSON manifest it just downloaded and grab the necessary URLs/files.

tarasglek commented:
The problem was plainly obvious in error log, though I missed it on first pass too. I added non-interactive flag so we don't run into this stupid ux issue again.

```
[eget.wasm] using stderr file /tmp/eget.stderr_6da829c9-2026-4ff0-8211-2f105a61f001.txt
[eget.wasm] Initiating compilation for WASM module: /home/taras/Documents/eget.wasm/eget.wasm
[eget.wasm] Compiled WASM module: /home/taras/Documents/eget.wasm/eget.wasm
[eget.wasm] Starting WASM execution...
[eget.wasm] WASM execution finished with exit code: 1
[eget.wasm] exited with code 1: 2 matches found: please select manually
Multiple candidates found, but interactive mode is disabled.
Please select one of the following candidates:
  - sops-v3.10.2.linux.amd64
  - sops-v3.10.2.linux.amd64.spdx.sbom.json
Use the --asset flag to select one of the above candidates.
```

humphd commented:
Amazing, thank you for debugging this.

I pulled your changes and did some more testing, but I can't get it to actually download the files, even if I use `--all` and `--files '*'`:

```js
import { eget } from "./eget.js";

async function main() {
  await eget('getsops/sops', { verbose: true, all: true, file: "'*'" });
  console.log('done')
}

main();
```

```sh
$ node test.js
[eget.wasm] Running eget with args: --non-interactive --system darwin/arm64 --file '*' --all getsops/sops
(node:7984) ExperimentalWarning: WASI is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
[eget.wasm] using stderr file /var/folders/82/9ymnd56s1g17vffjb280x32c0000gn/T/eget.stderr_0fe77f09-f8f0-4083-b122-e7e3d03cec22.txt
[eget.wasm] Initiating compilation for WASM module: /Users/humphd/repos/eget.wasm/eget.wasm
[eget.wasm] Compiled WASM module: /Users/humphd/repos/eget.wasm/eget.wasm
[eget.wasm] Starting WASM execution...
[eget.wasm] WASM execution finished with exit code: 1
[eget.wasm] exited with code 1: open /tmp/https/api.github.com/repos/getsops/sops/releases/latest: No such file or directory
[eget.wasm] Attempt 1: eget failed - open /tmp/https/api.github.com/repos/getsops/sops/releases/latest: No such file or directory
[eget.wasm] Downloading: https://api.github.com/repos/getsops/sops/releases/latest to /Users/humphd/repos/eget.wasm/tmp/https/api.github.com/repos/getsops/sops/releases/latest
[eget.wasm] Saved to: /Users/humphd/repos/eget.wasm/tmp/https/api.github.com/repos/getsops/sops/releases/latest
[eget.wasm] using stderr file /var/folders/82/9ymnd56s1g17vffjb280x32c0000gn/T/eget.stderr_ee9b6784-cbc9-481c-ab54-fcc181446a94.txt
[eget.wasm] Using cached WASM compilation promise
[eget.wasm] Starting WASM execution...
[eget.wasm] WASM execution finished with exit code: 1
[eget.wasm] exited with code 1: 2 matches found: please select manually
Multiple candidates found, but interactive mode is disabled.
Please select one of the following candidates:
  - sops-v3.10.2.darwin.arm64
  - sops-v3.10.2.darwin.arm64.spdx.sbom.json
Use the --asset flag to select one of the above candidates.
[eget.wasm] Attempt 2: eget failed - 2 matches found: please select manually
Multiple candidates found, but interactive mode is disabled.
Please select one of the following candidates:
  - sops-v3.10.2.darwin.arm64
  - sops-v3.10.2.darwin.arm64.spdx.sbom.json
Use the --asset flag to select one of the above candidates.
[eget.wasm] eget failed with no recovery URL. Error: 2 matches found: please select manually
Multiple candidates found, but interactive mode is disabled.
Please select one of the following candidates:
  - sops-v3.10.2.darwin.arm64
  - sops-v3.10.2.darwin.arm64.spdx.sbom.json
Use the --asset flag to select one of the above candidates.
[eget.wasm] Cleaned up temporary files in /Users/humphd/repos/eget.wasm/tmp
done
```

It shouldn't be necessary to specify the specific `asset` by name.

humphd commented:
Getting closer. If I explicitly tell it to ignore the `json` candidate, it downloads; but the call to `stat .` is failing:

```js
import { eget } from "./eget.js";

async function main() {
  await eget('getsops/sops', { quiet: true, asset: '^json', verbose: true });
  console.log('done')
}

main();
```

```sh
$ node test
[eget.wasm] Running eget with args: --non-interactive --system darwin/arm64 --quiet --asset ^json getsops/sops
(node:48117) ExperimentalWarning: WASI is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
[eget.wasm] using stderr file /var/folders/82/9ymnd56s1g17vffjb280x32c0000gn/T/eget.stderr_85fa0e5f-a5c8-464a-8d63-69f34cd46d52.txt
[eget.wasm] Initiating compilation for WASM module: /Users/humphd/repos/eget.wasm/eget.wasm
[eget.wasm] Compiled WASM module: /Users/humphd/repos/eget.wasm/eget.wasm
[eget.wasm] Starting WASM execution...
[eget.wasm] WASM execution finished with exit code: 1
[eget.wasm] exited with code 1: open /tmp/https/api.github.com/repos/getsops/sops/releases/latest: No such file or directory
[eget.wasm] Attempt 1: eget failed - open /tmp/https/api.github.com/repos/getsops/sops/releases/latest: No such file or directory
[eget.wasm] Downloading: https://api.github.com/repos/getsops/sops/releases/latest to /Users/humphd/repos/eget.wasm/tmp/https/api.github.com/repos/getsops/sops/releases/latest
[eget.wasm] Saved to: /Users/humphd/repos/eget.wasm/tmp/https/api.github.com/repos/getsops/sops/releases/latest
[eget.wasm] using stderr file /var/folders/82/9ymnd56s1g17vffjb280x32c0000gn/T/eget.stderr_d45c0cc3-9968-4811-8dc8-bafb716313c1.txt
[eget.wasm] Using cached WASM compilation promise
[eget.wasm] Starting WASM execution...
[eget.wasm] WASM execution finished with exit code: 1
[eget.wasm] exited with code 1: open /tmp/https/github.com/getsops/sops/releases/download/v3.10.2/sops-v3.10.2.darwin.arm64: No such file or directory
[eget.wasm] Attempt 2: eget failed - open /tmp/https/github.com/getsops/sops/releases/download/v3.10.2/sops-v3.10.2.darwin.arm64: No such file or directory
[eget.wasm] Downloading: https://github.com/getsops/sops/releases/download/v3.10.2/sops-v3.10.2.darwin.arm64 to /Users/humphd/repos/eget.wasm/tmp/https/github.com/getsops/sops/releases/download/v3.10.2/sops-v3.10.2.darwin.arm64
[eget.wasm] Saved to: /Users/humphd/repos/eget.wasm/tmp/https/github.com/getsops/sops/releases/download/v3.10.2/sops-v3.10.2.darwin.arm64
[eget.wasm] using stderr file /var/folders/82/9ymnd56s1g17vffjb280x32c0000gn/T/eget.stderr_46e175e9-41e0-4381-8d98-3120f06ce639.txt
[eget.wasm] Using cached WASM compilation promise
[eget.wasm] Starting WASM execution...
[eget.wasm] WASM execution finished with exit code: 1
[eget.wasm] exited with code 1: stat .: No such file or directory
[eget.wasm] Attempt 3: eget failed - stat .: No such file or directory
[eget.wasm] eget failed with no recovery URL. Error: stat .: No such file or directory
[eget.wasm] Cleaned up temporary files in /Users/humphd/repos/eget.wasm/tmp
done
```

So within the WASM filesystem, it doesn't like `.`.  I tried setting `--to` but that didn't seem to work either.

humphd commented:
OK, I built on your `to: '/'` idea and have it working.  My idea is that we always have the WASM code use `/` in the mapped WASI filesystem.  I achieve this by handling the `--to` and file moving logic in JS vs. WASM, and passing `--to '/'` into WASM so it always uses the root.

I need to update tests, but the `repro.js` is working nicely.  I added a second example that uses an explicit `to` to make sure that works.

humphd commented:
Added more `repo.js` examples.  You can run it now with `node repo good` or `node repo bad` or `node repo all` based on the get test cases.  Lots working, some edge cases I don't understand.

humphd commented:
Added an optional `onProgress` property that you can use to render some kind of progress update on the download(s), similar to what the Go version does:

```sh
Downloading sops-v3.10.2.darwin.arm64 | ████████████████████████████████████████ | 44.2 MB / 44.2 MB (100%)
```

See the `sops` example in `repro.js`, which uses it.

humphd commented:
Did some clean-up on the TS side of this:

- Moved type definitions to `eget.d.ts`, with `imports` in the JSDoc,  which is nicer to use vs. doing it all in JSDoc
- Moved to `jsconfig.json` vs. `tsconfig.json` and got full typechecking, fixed type errors with more annotations

humphd commented:
As far as I can tell, this is basically done and working for most cases.  I need to:

- fix tests so they pass (I'll use stuff from repro.js)
- update docs to match new types (e.g., onProgress)

Once that's in, I'll land and do the publish to npm bits in a follow-up.  We can fix bugs/add things in follow-ups.

humphd commented:
OK, this is done and I'm ready to land.

Before I publish this to NPM, I'll check in with the upstream repo to see if they want any involvement.

humphd commented:
https://github.com/zyedidia/eget/issues/131
